adobe/himl
# add support for custom merge strategy

## Description
Adding custom merge strategy to himl so default behavior can be customized. 

Currently ConfigProcessor builds a Merger using 

`merger = Merger([(list, ["append"]), (dict, ["merge"])], ["override"], ["override"])`

We want to add support for custom merge strategies.  i.e. for a list of items, we want to be able to override the default values, instead of default append behavior. Also we would like to be able to remove certain default values from lower nodes in the hierarchy.

We can change ConfigProcessor and ConfigGenerator init method signature so those optional merge strategies can be passed in, and send to deepmerge Merger object.

## Related Issue






## Motivation and Context


We want to define certain default values for overall hierarchy, but also allows flexibility to override or remove default values by lower nodes if necessary. Currently the merging policy/strategy is hardcoded in ConfigGenerator. We'd like to propose to change that as an optional parameter so user can pass in customized functions if they have that need.

## How Has This Been Tested?




1. Modified main.py so an example custom merge strategy as a function is defined, and this function is passed in as an optional parameter to ConfigProcessor
2. Modify default.yaml and cluster.yaml to provide an example on how to use the custom merge strategy. User can define different unique keys for each record to distinguish records. Also a "remove" flag is provided in the example to showcase how to delete a default value.
3. Run `himl examples/complex/env=dev/region=us-east-1/cluster=cluster2 ` to test expected behavior: 
 * records with same 'id' field are overridden 
 * record with 'remove' flag is removed
4. Run the same command against other folder such as `himl examples/complex/env=dev/region=us-east-1/cluster=cluster1` to make sure default behavior is not changed if no 'id' field is defined.


## Screenshots (if appropriate):

## Types of changes




## Checklist:
